### ISSUE TITLE
how to make train,test,val pt ??

### ISSUE THREAD

**Issue by youngseoh** [cyye001/Con-CDVAE#7]
how to make train, test, val pt???
I have obtained the MP20 dataset through an API, but I don't know how to convert it into a .pt file

**Comment by cyye001** [comment]
When you use the dataset for the first time, the code will read the CSV file and store the processed data in the same path as xxx.pt files.

**Comment by youngseoh** [comment]
Thank you for your response; the issue has been resolved. However, when I run the model with the material project data I downloaded, the atom_type_probs keeps generating negative values, which is disrupting the training process. If possible, could you personally share the CSV or .pt files you used for train, test, and validation in the material project? Here’s my email: youngseoh6@gmail.com. I would greatly appreciate any help. Thank you very much!

**Comment by cyye001** [comment]
This error may not be caused by the dataset. You could try using the dataset provided by Tian Xie—I initially used this dataset for testing as well. 
https://github.com/txie-93/cdvae/tree/main/data/mp_20

If possible, could you show me the detailed error message you’re encountering?

**Comment by youngseoh** [comment]
I also downloaded the train, test, and val datasets from [this link](https://github.com/txie-93/cdvae/tree/main/data/mp_20) and used them in my project. However, when I try to run the training, the following error occurs:
![image](https://github.com/user-attachments/assets/fbdcf2c6-f4ba-4d64-bbfc-5288f3c8351c)

**Comment by cyye001** [comment]
In fact, what you’re encountering is likely not a negative value but rather a NaN.

I think it is caused by exploding gradients. You can avoid this by setting gradient clipping and reducing the learning rate.
If you want to set gradient clipping, you can set like this:
python concdvae/run.py train=new data=mptest expname=test model=vae_mp_format_gap train.PT_train.clip_grad_norm=0.001 
If you want to change learning rate, you can find it in conf/optim/default.yaml.

By the way, another study (cond-CDVAE) found that adding a tanh function after the encoder can effectively improve this situation. For example you can try add 'energy = torch.tanh(energy)' before 'return energy' in line 426 of concdvae/pl_modules/gnn.py.

**Comment by youngseoh** [comment]
You're welcome! I'm glad the issue was resolved. 
I have one more question. In the [CDVAE repository](https://github.com/txie-93/cdvae/tree/main/data/mp_20), instead of downloading the mp_20 dataset, I am trying to retrieve data directly from the Materials Project API. However, variables such as band_gap, e_above_hull, CS_type, BG_type, and FM_type do not exist as fields in the API. How were these obtained? Could it be that they were processed afterward?

**Comment by cyye001** [comment]
The band_gap and e_above_hull can be downloaded with the key word "band_gap" and "energy_above_hull" by using the API. CS_type, BG_type, and FM_type represent types of crystal system, band gap and formation energy, respectively, which are the labels I annotated myself. You can determine the crystal system based on the space group. The space group and formation energy can also be downloaded by using the API. BG_type and FM_type were defined by me while testing the code, so you don’t need to worry about them.
